packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaFormReducer.ts

```diff
@@ -0,0 +1,500 @@
+) {
+  const { columnModelIndex, jsonSubColumnModelIndex, newColumnType } = action
+  let newColumnModelValue: ColumnModelFormData | JsonSubColumnModelFormData
+
+  // Create a copy of the selected column model
+  if (
+    prevState &&
+    prevState[columnModelIndex] &&
+    prevState[columnModelIndex].jsonSubColumns &&
+    jsonSubColumnModelIndex !== undefined
+  ) {
+    newColumnModelValue = cloneDeep(
+      prevState[columnModelIndex].jsonSubColumns![jsonSubColumnModelIndex],
+    )
+  } else {
+    newColumnModelValue = cloneDeep(prevState[columnModelIndex])
+  }
+
+  // Update the column model. Remove fields that no longer make sense for the new column type
+  newColumnModelValue.columnType = newColumnType
+  if (!canHaveSize(newColumnType) && 'maximumSize' in newColumnModelValue) {
+    delete newColumnModelValue.maximumSize
+  }
+  if (
+    !canHaveMaxListLength(newColumnType) &&
+    'maximumListLength' in newColumnModelValue
+  ) {
+    delete newColumnModelValue.maximumListLength
+  }
+  if (
+    !canHaveRestrictedValues(newColumnType, !!jsonSubColumnModelIndex) &&
+    'enumValues' in newColumnModelValue
+  ) {
+    delete newColumnModelValue.enumValues
+  }
+
+  const allowedFacetTypes = configureFacetsForType(
+    newColumnType,
+    !!jsonSubColumnModelIndex,
+  )
+  if (
+    'facetType' in newColumnModelValue &&
+    (allowedFacetTypes === null ||
+      !allowedFacetTypes.includes(newColumnModelValue.facetType))
+  ) {
+    delete newColumnModelValue.facetType
+  }
+
+  if (
+    'jsonSubColumns' in newColumnModelValue &&
+    newColumnType !== ColumnTypeEnum.JSON
+  ) {
+    delete newColumnModelValue.jsonSubColumns
+  }
+
+  // Replace the value
+  if (
+    prevState &&
+    prevState[columnModelIndex] &&
+    prevState[columnModelIndex].jsonSubColumns &&
+    jsonSubColumnModelIndex !== undefined
+  ) {
+    prevState[columnModelIndex].jsonSubColumns![jsonSubColumnModelIndex] =
+      newColumnModelValue as JsonSubColumnModelFormData
+  } else {
+    prevState[columnModelIndex] = newColumnModelValue as ColumnModelFormData
+  }
+}
+
+function toggleSelectAll(prevState: ColumnModelFormData[]) {
+  const allSelected = getIsAllSelected(prevState)
+  if (allSelected) {
+    return prevState.map(cm => ({
+      ...cm,
+      jsonSubColumns: cm.jsonSubColumns?.map(jsc => ({
+        ...jsc,
+        isSelected: false,
+      })),
+      isSelected: false,
+    }))
+  } else {
+    return prevState.map(cm => ({
+      ...cm,
+      jsonSubColumns: cm.jsonSubColumns?.map(jsc => ({
+        ...jsc,
+        isSelected: true,
+      })),
+      isSelected: true,
+    }))
+  }
+}
+
+function setColumnModelValue(
+  action: {
+    type: 'setColumnModelValue'
+    columnModelIndex: number
+    jsonSubColumnModelIndex?: number
+    value: ColumnModelFormData | JsonSubColumnModelFormData
+  },
+  prevState: ColumnModelFormData[],
+) {
+  const { columnModelIndex, jsonSubColumnModelIndex, value } = action
+  if (
```

Comment:
Should this be something like this (same comment for a subset of other utility functions)?
if (prevState && prevState[columnModelIndex]) {
  if (prevState[columnModelIndex].jsonSubColumns &&
    jsonSubColumnModelIndex !== undefined) { ...}
  else {...}
}

Reply:
Good idea, that would definitely be safer